api/src/test/java/org/pmiops/workbench/audit/adapters/WorkspaceAuditAdapterServiceTest.java

```diff
@@ -0,0 +1,129 @@
+import org.pmiops.workbench.audit.ActionAuditEvent;
+import org.pmiops.workbench.audit.ActionAuditService;
+import org.pmiops.workbench.audit.ActionType;
+import org.pmiops.workbench.db.model.User;
+import org.pmiops.workbench.model.DataAccessLevel;
+import org.pmiops.workbench.model.ResearchPurpose;
+import org.pmiops.workbench.model.Workspace;
+import org.pmiops.workbench.workspaces.WorkspaceConversionUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.TestConfiguration;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.test.context.junit4.SpringRunner;
+
+@RunWith(SpringRunner.class)
+public class WorkspaceAuditAdapterServiceTest {
+
+  private static final long WORKSPACE_1_DB_ID = 101L;
+  private WorkspaceAuditAdapterService workspaceAuditAdapterService;
+  private Workspace workspace1;
+  private User user1;
+  private org.pmiops.workbench.db.model.Workspace dbWorkspace1;
+
+  @Mock private Provider<User> mockUserProvider;
+  @Autowired private ActionAuditService mockActionAuditService;
+  @Captor private ArgumentCaptor<Collection<ActionAuditEvent>> eventListCaptor;
+  @Captor private ArgumentCaptor<ActionAuditEvent> eventCaptor;
+
+  @TestConfiguration
+  @MockBean(value = {ActionAuditService.class})
+  static class Configuration {}
+
+  @Before
+  public void setUp() {
+    user1 = new User();
+    user1.setUserId(101L);
+    user1.setEmail("fflinstone@slate.com");
+    user1.setGivenName("Fred");
+    user1.setFamilyName("Flintstone");
+    doReturn(user1).when(mockUserProvider).get();
+    workspaceAuditAdapterService =
+        new WorkspaceAuditAdapterServiceImpl(mockUserProvider, mockActionAuditService);
+
+    final ResearchPurpose researchPurpose1 = new ResearchPurpose();
+    researchPurpose1.setIntendedStudy("stubbed toes");
+    researchPurpose1.setAdditionalNotes("I really like the cloud.");
+    final long now = System.currentTimeMillis();
+
+    workspace1 = new Workspace();
+    workspace1.setName("Workspace 1");
+    workspace1.setId("fc-id-1");
+    workspace1.setNamespace("aou-rw-local1-c4be869a");
+    workspace1.setCreator("user@fake-research-aou.org");
+    workspace1.setCdrVersionId("1");
+    workspace1.setResearchPurpose(researchPurpose1);
+    workspace1.setCreationTime(now);
+    workspace1.setLastModifiedTime(now);
+    workspace1.setEtag("etag_1");
+    workspace1.setDataAccessLevel(DataAccessLevel.REGISTERED);
+    workspace1.setPublished(false);
+
+    dbWorkspace1 = WorkspaceConversionUtils.toDbWorkspace(workspace1);
+    dbWorkspace1.setWorkspaceId(WORKSPACE_1_DB_ID);
+    dbWorkspace1.setLastAccessedTime(new Timestamp(now));
+    dbWorkspace1.setLastModifiedTime(new Timestamp(now));
+    dbWorkspace1.setCreationTime(new Timestamp(now));
+  }
+
+  @Test
+  public void testFiresCreateWorkspaceEvents() {
+    workspaceAuditAdapterService.fireCreateAction(workspace1, WORKSPACE_1_DB_ID);
```

Comment:
can we test one level higher by creating workspaces through our controllers? I think that adds a bit more coverage .

Reply:
Adding a few tests to `WorkspacesControllerTest` works too. I just want to make sure the whole flow is captured through our tests. It's fine if it's split up.

Reply:
All there was to do was verify we're calling the adapter service.